src/Features/CSharp/Portable/UseIndexOperator/CSharpUseIndexOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+    /// order to do this, the type must look 'indexable'.  Meaning, it must have an int-returning
+    /// property called 'Length' or 'Count', and it must have both an int-indexer, and a
+    /// System.Index indexer.
+    ///
+    /// It is assumed that if the type follows this shape that it is well behaved and that this
+    /// transformation will preserve semantics.  If this assumption is not good in practice, we
+    /// could always limit the feature to only work on a whitelist of known safe types.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp), Shared]
+    internal partial class CSharpUseIndexOperatorDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUseIndexOperatorDiagnosticAnalyzer() 
+            : base(IDEDiagnosticIds.UseIndexOperatorDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Use_index_operator), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(FeaturesResources.Indexing_can_be_simplified), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                // We're going to be checking every property-reference in the compilation. Cache
+                // information we compute in this TypeChecker object so we don't have to continually
+                // recompute it.
+                var typeChecker = new TypeChecker(startContext.Compilation);
+                context.RegisterOperationAction(
+                    c => AnalyzePropertyReference(c, typeChecker),
+                    OperationKind.PropertyReference);
+            });
+        }
+
+        private void AnalyzePropertyReference(
+            OperationAnalysisContext context, TypeChecker typeChecker)
+        {
+            var cancellationToken = context.CancellationToken;
+            var propertyReference = (IPropertyReferenceOperation)context.Operation;
+            var property = propertyReference.Property;
+
+            // Only analyze indexer calls.
+            if (!property.IsIndexer)
+            {
+                return;
+            }
+
+            // Make sure we're actually on something like `s[...]`.
+            var elementAccess = propertyReference.Syntax as ElementAccessExpressionSyntax;
+            if (elementAccess == null)
+            {
+                return;
+            }
+
+            // Only supported on C# 8 and above.
+            var syntaxTree = elementAccess.SyntaxTree;
+            var parseOptions = (CSharpParseOptions)syntaxTree.Options;
+            //if (parseOptions.LanguageVersion < LanguageVersion.CSharp8)
+            //{
+            //    return;
+            //}
```

Comment:
can't actually set hte IDE to C# 8 mode yet.  So this is disabled in the meantime for experimentation purposes.